simulation/traffic_simulator/src/traffic_lights/traffic_light_publisher.cpp

```diff
@@ -0,0 +1,89 @@
+  const rclcpp::Time & current_ros_time,
+  const simulation_api_schema::UpdateTrafficLightsRequest & request) -> void
+{
+  autoware_auto_perception_msgs::msg::TrafficSignalArray msg;
+  using TrafficLightType = autoware_auto_perception_msgs::msg::TrafficSignal;
+  msg.header.frame_id = "camera_link";  // DIRTY HACK!!!
+  msg.header.stamp = current_ros_time;
+  for (const auto & traffic_light : request.states()) {
+    TrafficLightType traffic_light_msg;
+    traffic_light_msg.map_primitive_id = traffic_light.id();
+    for (auto bulb_status : traffic_light.traffic_light_status()) {
+      using TrafficLightBulbType = TrafficLightType::_lights_type::value_type;
+      TrafficLightBulbType light_bulb_msg;
+      simulation_interface::toMsg<TrafficLightBulbType>(bulb_status, light_bulb_msg);
+      traffic_light_msg.lights.push_back(light_bulb_msg);
+    }
+    msg.signals.push_back(traffic_light_msg);
+  }
+  traffic_light_state_array_publisher_->publish(msg);
+}
+
+template <>
+auto TrafficLightPublisher<autoware_perception_msgs::msg::TrafficSignalArray>::publish(
+  const rclcpp::Time & current_ros_time,
+  const simulation_api_schema::UpdateTrafficLightsRequest & request) -> void
+{
+  assert(hdmap_utils_ != nullptr);
+  // cache store for conversion from way_id to relation_id
+  static std::unordered_map<std::int64_t, std::vector<std::int64_t>> id_conversion_cache_map;
+
+  autoware_perception_msgs::msg::TrafficSignalArray msg;
+  msg.stamp = current_ros_time;
+  for (const auto & traffic_light : request.states()) {
+    auto relation_ids = [&]() -> std::vector<std::int64_t> {
+      if (auto cache = id_conversion_cache_map.find(traffic_light.id());
+          cache != id_conversion_cache_map.end()) {
+        // use cached relation_id if the conversion is already cached
+        return cache->second;
+      } else {
+        auto relation_ids = hdmap_utils_->getTrafficLightRelationIDsFromWayID(traffic_light.id());
+        // cache the conversion
+        id_conversion_cache_map[traffic_light.id()] = relation_ids;
+        return relation_ids;
+      }
+    }();
+
+    for (auto relation_id : relation_ids) {
+      // skip if the traffic light has no bulbs
+      if (not traffic_light.traffic_light_status().empty()) {
+        using TrafficLightType = autoware_perception_msgs::msg::TrafficSignal;
+        TrafficLightType traffic_light_msg;
+        traffic_light_msg.traffic_signal_id = relation_id;
+
+        for (auto bulb_status : traffic_light.traffic_light_status()) {
+          using TrafficLightBulbType =
+            autoware_perception_msgs::msg::TrafficSignal::_elements_type::value_type;
+          TrafficLightBulbType light_bulb_msg;
+          simulation_interface::toMsg<TrafficLightBulbType>(bulb_status, light_bulb_msg);
+          traffic_light_msg.elements.push_back(light_bulb_msg);
+        }
+        msg.signals.push_back(traffic_light_msg);
+      }
+    }
+  }
+  traffic_light_state_array_publisher_->publish(msg);
+}
```

Comment:
Did you implement caching because this function was found to be slow based on measurements based on comparable metrics? If this is not the case, the function is very likely performing "premature optimization". Remove any caching code and keep this function simple.

Even if caching were necessary, I don't think this function's caching process itself would be efficient because there are too many unnecessary copies.

Reply:
This caching seemed to apply to "premature optimization", so I deleted it in 94e71837d836281f851d0feef63df375adb991e4.